### PR TITLE
Added PILToTensor and ConvertImageDtype classes in reference scripts

### PR DESCRIPTION
**Pull request by prabhat00155**
Detection:
```
srun -p train --cpus-per-task=16 -t 10:00:00 --gpus-per-node=8 python -m torch.distributed.launch --nproc_per_node=8 --use_env train.py --model fasterrcnn_resnet50_fpn --output-dir logs/to_tensor/ --pretrained --test-only
```
Without these changes:
```
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.369
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.585
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.396
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.212
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.403
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.482
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.307
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.485
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.509
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.317
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.544
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.649
```
With these changes:
```
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.369
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.585
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.396
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.212
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.403
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.482
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.307
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.485
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.509
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.317
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.544
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.649
```
Segmentation:
```
srun -p train --cpus-per-task=16 -t 10:00:00 --gpus-per-node=8 python -m torch.distributed.launch --nproc_per_node=8 --use_env train.py --model fcn_resnet50 --output-dir logs/to_tensor/ --pretrained --test-only
```
Without these changes:
```
global correct: 91.4
average row correct: ['94.0', '80.9', '73.9', '72.3', '49.9', '56.9', '81.0', '66.5', '88.6', '43.2', '75.7', '67.6', '80.9', '84.3', '81.2', '90.9', '49.0', '80.4', '61.2', '77.7', '69.0']
IoU: ['90.3', '72.6', '61.2', '61.6', '42.7', '46.9', '73.0', '53.7', '80.3', '33.6', '64.1', '33.9', '61.2', '72.0', '73.6', '80.7', '29.1', '63.5', '48.2', '71.2', '56.3']
mean IoU: 60.5
```
With these changes:
```
Test: Total time: 0:00:49
global correct: 91.4
average row correct: ['94.0', '80.9', '73.9', '72.3', '49.9', '56.9', '81.0', '66.5', '88.6', '43.2', '75.7', '67.6', '80.9', '84.3', '81.2', '90.9', '49.0', '80.4', '61.2', '77.7', '69.0']
IoU: ['90.3', '72.6', '61.2', '61.6', '42.7', '46.9', '73.0', '53.7', '80.3', '33.6', '64.1', '33.9', '61.2', '72.0', '73.6', '80.7', '29.1', '63.5', '48.2', '71.2', '56.3']
mean IoU: 60.5
```

cc @vfdev-5 @datumbox

### CODE REVIEW

File: references/detection/presets.py [hunk 1]
```diff
@@ -6,21 +6,24 @@ def __init__(self, data_augmentation, hflip_prob=0.5, mean=(123., 117., 104.)):
         if data_augmentation == 'hflip':
             self.transforms = T.Compose([
                 T.RandomHorizontalFlip(p=hflip_prob),
-                T.ToTensor(),
+                T.PILToTensor(),
+                T.ConvertImageDtype(),
```
Comment (datumbox) [inline references/detection/presets.py:10]: What are your thoughts on provideing the same API as the equivalent class on transforms?
```suggestion
                T.ConvertImageDtype(dtype=torch.float),
```

File: references/detection/transforms.py [hunk 2]
```diff
@@ -52,6 +52,20 @@ def forward(self, image: Tensor,
...
 
 
+class PILToTensor(nn.Module):
+    def forward(self, image: Tensor,
+                target: Optional[Dict[str, Tensor]] = None) -> Tuple[Tensor, Optional[Dict[str, Tensor]]]:
+        image = F.pil_to_tensor(image)
+        return image, target
+
+
+class ConvertImageDtype(nn.Module):
```
Comment (datumbox) [inline references/detection/transforms.py:62]: How about something like this?
```python
    def __init__(self, dtype: torch.dtype) -> None:
        super().__init__()
        self.dtype = dtype
```

File: references/segmentation/transforms.py [hunk 3]
```diff
@@ -75,9 +74,15 @@ def __call__(self, image, target):
         return image, target
 
 
-class ToTensor(object):
+class PILToTensor:
     def __call__(self, image, target):
         image = F.pil_to_tensor(image)
+        target = torch.as_tensor(np.array(target), dtype=torch.int64)
```
Comment (datumbox) [inline references/segmentation/transforms.py:80]: What is the type of target passed to this method? Can we avoid using numpy here?

**Review by datumbox — COMMENTED** [review]
Thanks @prabhat00155, the numbers look OK though it would be good to provide also runs on the latest main to confirm.

For example on our docs we report that the mAP of the `Faster R-CNN ResNet-50 FPN` model should e 37.0 not 36.9 that you got but I believe this is the result of a separate accuracy deterioration and not related to your changes. By having runs on the latest main you can prove that this PR is not to blame.

Finally below I left some comments for your consideration, let me know what you think:

File: references/detection/transforms.py (hunk 2 above)
Comment (prabhat00155) [reply references/detection/transforms.py:62]: Makes sense. Should we have a default value for dtype(torch.float)?

Comment (datumbox) [reply references/detection/transforms.py:62]: I was proposing to completely align it with the actual `ConvertImageDtype` of TorchVision. I believe it doesn't set one. I think aligning them will make it super useful to switch to the new Transformations API that Francisco and Phillip are working on. That's the idea at least. Let me know if you agree.

Comment (prabhat00155) [reply references/detection/transforms.py:62]: Okay, sounds good.

File: references/segmentation/transforms.py (hunk 3 above)
Comment (prabhat00155) [reply references/segmentation/transforms.py]: Done. I haven't been able to test this yet as I am again getting error on aws cluster. I'll update here once I have completed the training run.

**Review by datumbox — APPROVED** [review]
LGTM, thanks!

File: references/segmentation/transforms.py (hunk 3 above)
Comment (prabhat00155) [reply references/segmentation/transforms.py]: Seems to be Image type. Without `np.array()` it throws this error: `TypeError: an integer is required (got type Image)`

File: references/segmentation/transforms.py [hunk 4]
```diff
@@ -75,14 +74,23 @@ def __call__(self, image, target):
...
         return image, target
 
 
+class ConvertImageDtype:
+    def __init__(self, dtype):
+        self.dtype = dtype
+
+    def __call__(self, image, target):
+        image = F.convert_image_dtype(image, self.dtype)
+        target = torch.as_tensor(target, dtype=torch.int64)
```
Comment (datumbox) [inline references/segmentation/transforms.py:90]: Why don't we let target unchanged here? Aka straight return it without modifications?

Comment (prabhat00155) [reply references/segmentation/transforms.py]: Yes, you are right. This is not needed. Thanks!

**Pull request merged**